### ISSUE TITLE
[MacOS] Close a invisible window cause segment fault.

### ISSUE THREAD

**Issue by kaiwk** [rust-windowing/winit#1863]
The code is modified from `examples/multiwindow.rs`, and reproduction is simple:

1. press 'a' will add an invisible window;
2. press 'x' will remove all windows except current window (which receives key events), then a segment fault throwed.

```rust
use std::collections::HashMap;

use simple_logger::SimpleLogger;
use winit::{
    event::{ElementState, Event, KeyboardInput, VirtualKeyCode, WindowEvent},
    event_loop::{ControlFlow, EventLoop},
    window::Window,
};

fn main() {
    SimpleLogger::new().init().unwrap();
    let event_loop = EventLoop::new();

    let mut windows = HashMap::new();
    let window = Window::new(&event_loop).unwrap();
    windows.insert(window.id(), window);

    event_loop.run(move |event, event_loop, control_flow| {
        *control_flow = ControlFlow::Wait;

        match event {
            Event::WindowEvent { event, window_id } => {
                match event {
                    WindowEvent::CloseRequested => {
                        println!("Window {:?} has received the signal to close", window_id);

                        // This drops the window, causing it to close.
                        windows.remove(&window_id);

                        if windows.is_empty() {
                            *control_flow = ControlFlow::Exit;
                        }
                    }
                    WindowEvent::KeyboardInput {
                        input:
                            KeyboardInput {
                                state: ElementState::Pressed,
                                virtual_keycode: Some(VirtualKeyCode::A),
                                ..
                            },
                        ..
                    } => {
                        let window = Window::new(&event_loop).unwrap();
                        window.set_visible(false);
                        windows.insert(window.id(), window);
                    }
                    WindowEvent::KeyboardInput {
                        input:
                            KeyboardInput {
                                state: ElementState::Pressed,
                                virtual_keycode: Some(VirtualKeyCode::X),
                                ..
                            },
                        ..
                    } => {
                        windows.retain(|id, _| *id == window_id);
                    }
                    _ => (),
                }
            }
            _ => (),
        }
    })
}
```

I guess it's related to #158?

**Comment by kaiwk** [comment]
From the [nswindow doc](https://developer.apple.com/documentation/appkit/nswindow/1419662-close?language=objc), even for invisible window, 'nswindow.close()' should post `NSWindowWillCloseNotification`, but looks like [window_will_close](https://github.com/rust-windowing/winit/blob/dd32ace9ab4179cdbbccdc191d2bc7a513569c10/src/platform_impl/macos/window_delegate.rs#L273) is not triggtered from the trace log.

**Comment by kaiwk** [comment]
It seems window delegate drops before `UnownedWindow` somehow, one quick dirty fix is [here](https://github.com/kaiwk/winit/blob/d839f868d23de02cbf2a533854b6e2960dce1c3f/src/platform_impl/macos/window.rs#L1157).

**Comment by ArturKovacs** [comment]
Please check if this is fixed on the master currently. I believe this was fixed by #1874.

**Comment by kaiwk** [comment]
Thanks, I tested it and it works.
